src/Utilities/functions.jl

```diff
@@ -215,27 +250,52 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
 The canonical representation of `ScalarAffineFunction([y, x, z, x, z], [2, 1, 3, -2, -3], 5)` is `ScalarAffineFunction([x, y], [-1, 2], 5)`.
 
 """
-function canonical(f::Union{SAF, VAF})
-    sorted_terms = sort(f.terms, by = termindices)
-    terms = eltype(f.terms)[]
-    for t in sorted_terms
-        tnot0 = !iszero(coefficient(t))
-        if !isempty(terms) && termindices(t) == termindices(last(terms)) && tnot0
-            terms[end] = unsafe_add(t, last(terms))
-        elseif tnot0
-            if !isempty(terms) && iszero(coefficient(last(terms)))
-                terms[end] = t
+canonical(f::Union{SAF, VAF}) = canonicalize!(copy(f))
+
+"""
+    canonicalize!(f::Union{ScalarAffineFunction, VectorAffineFunction})
+
+Convert a function to canonical form in-place, without allocating a copy to hold the result.
+See [`canonical`](@ref).
+"""
+function canonicalize!(f::Union{SAF, VAF})
+    sort_and_compress!(f.terms, termindices, t -> !iszero(coefficient(t)), unsafe_add)
+    f
+end
+
+"""
+    sort_and_compress!(x::AbstractVector, by::Function, keep::Function, combine::Function)
+
+Sort the vector `x` in-place using `by` as the function from elements to comparable keys, then
+combine all entries for which `by(x[i]) == by(x[j])` using the function `x[i] = combine(x[i], x[j])`,
+and remove any entries for which `keep(x[i]) == false`. This may result in `x` being resized to
+a shorter length.
+"""
+function sort_and_compress!(x::AbstractVector, by, keep, combine)
+    if length(x) > 0
+        sort!(x, QuickSort, Base.Order.ord(isless, by, false, Base.Sort.Forward))
+        i1 = firstindex(x)
+        for i2 in eachindex(x)[2:end]
```

Comment:
Yeah, it's a generalization. The implementation is different though:

```julia
function unique(f::Callable, C)
    out = Vector{eltype(C)}()
    seen = Set()
    for x in C
        y = f(x)
        if !in(y, seen)
            push!(seen, y)
            push!(out, x)
        end
    end
    out
end
```

(on 0.6). On 0.7 there is also a `unique!(C)` (but interestingly there is no `unique!(f, C)`). And they still allocate and use a different algorithm.

Reply:
It's actually different anyway, as `unique` preserves the original order of appearance rather than sorting. So this is similar to `unique!(sort!(x))`, but I really don't think we need to go into that.